cmake/extensions.cmake

```diff
@@ -767,6 +774,148 @@ function(board_finalize_runner_args runner)
+              requires a revision. Please use: `-DBOARD=${BOARD}@<revision>`")
+    endif()
+  endif()
+
+  if(DEFINED BOARD_REV_HIGHEST_REVISION)
+    if(((BOARD_REV_FORMAT STREQUAL LETTER) AND
+        (BOARD_REVISION STRGREATER BOARD_REV_HIGHEST_REVISION)) OR
+       ((BOARD_REV_FORMAT MATCHES "^MAJOR\.MINOR\.PATCH$") AND
+        (BOARD_REVISION VERSION_GREATER BOARD_REV_HIGHEST_REVISION))
+    )
+      message(FATAL_ERROR "Board revision `${BOARD_REVISION}` greater than \
+              highest supported revision `${BOARD_REV_HIGHEST_REVISION}`. \
+	      Please specify a valid board revision.")
+    endif()
+  endif()
+
+  if(BOARD_REV_FORMAT STREQUAL LETTER)
+    set(revision_regex "([A-Z])")
+  elseif(BOARD_REV_FORMAT MATCHES "^MAJOR\.MINOR\.PATCH$")
+    set(revision_regex "((0|[1-9]+)(\.[0-9]+)(\.[0-9]+))")
+    # We allow loose <board>@<revision> typing on command line.
+    # so append missing zeroes.
+    if(BOARD_REVISION MATCHES "((0|[1-9]+)(\.[0-9]+)?(\.[0-9]+)?)")
+      if(NOT CMAKE_MATCH_3)
+        set(BOARD_REVISION ${BOARD_REVISION}.0)
+        set(BOARD_REVISION ${BOARD_REVISION} PARENT_SCOPE)
+      endif()
+      if(NOT CMAKE_MATCH_4)
+        set(BOARD_REVISION ${BOARD_REVISION}.0)
+        set(BOARD_REVISION ${BOARD_REVISION} PARENT_SCOPE)
+      endif()
+    endif()
+  else()
+    message(FATAL_ERROR "Invalid format specified for \
+    `zephyr_check_board_revision(FORMAT <LETTER | MAJOR.MINOR.PATCH>)`")
+  endif()
+
+  if(NOT (BOARD_REVISION MATCHES "^${revision_regex}$"))
+    message(FATAL_ERROR "Invalid revision format used for `${BOARD_REVISION}`. \
+            This board uses revision format: ${BOARD_REV_FORMAT}.")
+  endif()
+
+  string(REPLACE "." "_" underscore_revision_regex ${revision_regex})
+  set(file_revision_regex "${BOARD}_${underscore_revision_regex}.conf")
+  foreach(candidate ${revision_candidates})
+    if(${candidate} MATCHES "${file_revision_regex}")
+      string(REPLACE "_" "." FOUND_BOARD_REVISION ${CMAKE_MATCH_1})
+      if(${BOARD_REVISION} STREQUAL ${FOUND_BOARD_REVISION})
+        # Found exact match.
+        return()
+      endif()
+
+      if(NOT BOARD_REV_EXACT)
+        if((BOARD_REV_FORMAT MATCHES "^MAJOR\.MINOR\.PATCH$") AND
+           (${BOARD_REVISION} VERSION_GREATER_EQUAL ${FOUND_BOARD_REVISION}) AND
+           (${FOUND_BOARD_REVISION} VERSION_GREATER_EQUAL "${ACTIVE_BOARD_REVISION}")
+        )
+          set(ACTIVE_BOARD_REVISION ${FOUND_BOARD_REVISION})
+        elseif((BOARD_REV_FORMAT STREQUAL LETTER) AND
+               (${BOARD_REVISION} STRGREATER ${FOUND_BOARD_REVISION}) AND
+               (${FOUND_BOARD_REVISION} STRGREATER "${ACTIVE_BOARD_REVISION}")
+        )
+          set(ACTIVE_BOARD_REVISION ${FOUND_BOARD_REVISION})
+        endif()
+      endif()
+    endif()
+  endforeach()
+
+  if(BOARD_REV_EXACT OR NOT DEFINED ACTIVE_BOARD_REVISION)
+    message(FATAL_ERROR "Board revision `${BOARD_REVISION}` not found. \
```

Comment:
No, you would need to place an empty file.
Another thing you could do would be to move the config settings with changes out of defconfig and into specific revisions.

This also helps users to actual see what are revision specific settings.
So instead of:
```
File:plank_defconfig
CONFIG_FOO="old"

File:plank_1_0_0.conf
<empty>

File:plank_2_0_0.conf
CONFIG_FOO="new"
```

then do:
```
File:plank_defconfig
<empty>

File:plank_1_0_0.conf
CONFIG_FOO="old"

File:plank_2_0_0.conf
CONFIG_FOO="new"
```
and make it clear that `CONFIG_FOO` changes with different revisions.

Also the list of files helps everyone to see what revisions are valid.
I can do:
```
$ ls boards/arch/plank/*.conf
plank_9_0_0.conf
plank_10_0_0.conf
```
and I know I can overwrite values for revision 9 and 10 in `mysample`.

With your proposal, I cannot see that `9_0_0` is a valid revision to overwrite.